common/quotas/global/algorithm/scenarios_test.go

```diff
@@ -0,0 +1,478 @@
+					results.callRPS[2] += float64(r[2])
+					// find out how many would've been accepted and track it (uses previous-round target-RPS because limiters always lag by an update cycle)
+					track(&results.accepts[0], &results.rejects[0], r[0], results.targetRPS[0])
+					track(&results.accepts[1], &results.rejects[1], r[1], results.targetRPS[1])
+					track(&results.accepts[2], &results.rejects[2], r[2], results.targetRPS[2])
+					// update per-host RPS (running average in aggregator)
+					hostRPS[0] = average(hostRPS[0], r[0])
+					hostRPS[1] = average(hostRPS[1], r[1])
+					hostRPS[2] = average(hostRPS[2], r[2])
+
+					// update RPS that limiters allow.
+					// this is based on weighted data only, nothing directly from the current batch of data
+					totalRPS := hostRPS[0] + hostRPS[1] + hostRPS[2]
+					//                            vvvvvvvvvvvvvvvvvvvvv---- host's weight in cluster
+					results.targetRPS[0] = targetRPS * (hostRPS[0] / totalRPS)
+					results.targetRPS[1] = targetRPS * (hostRPS[1] / totalRPS)
+					results.targetRPS[2] = targetRPS * (hostRPS[2] / totalRPS)
+
+					t.Logf("round %d: "+
+						"calls: [%d, %d, %d], "+
+						"real: [%0.3f, %0.3f, %0.3f], "+
+						"rps: [%0.3f, %0.3f, %0.3f]",
+						ri,
+						r[0], r[1], r[2],
+						hostRPS[0], hostRPS[1], hostRPS[2],
+						results.targetRPS[0], results.targetRPS[1], results.targetRPS[2])
+				}
+
+				// and divide to get accept/reject-RPS (much more scale-agnostic)
+				results.accepts = divide(results.accepts, len(round))
+				results.rejects = divide(results.rejects, len(round))
+				results.callRPS = divide(results.callRPS, len(round))
+				results.cumulativeRPS = results.targetRPS[0] + results.targetRPS[1] + results.targetRPS[2]
+			}
+			return
+		}
+		t.Run("zeros", func(t *testing.T) {
+			// zeros are the worst-case scenario here, greatly favoring the non-zero values.
+			check(t, data{
+				targetRPS:     [3]float64{0.63, 3.57, 10.7}, // intermittent allows far too few, constant is allowed over 2x what it uses
+				callRPS:       [3]float64{1.25, 2.50, 5.00}, // all at or below per-host limit at all times, not just in aggregate
+				accepts:       [3]float64{0.00, 1.50, 5.00}, // this is the main problems with this algorithm: bad behavior when imbalanced but low usage.
+				rejects:       [3]float64{1.25, 1.00, 0.00},
+				cumulativeRPS: targetRPS, // but this is useful and simple: it always tries to enforce the "real" target, never over or under.
+			}, run(t, roundSchedule(0, 5)))
+		})
+		t.Run("low non-zeros", func(t *testing.T) {
+			// even 1 instead of 0 does quite a lot better, though still not great
+			check(t, data{
+				targetRPS:     [3]float64{2.21, 4.07, 8.72},
+				callRPS:       [3]float64{2.00, 3.00, 5.00},
+				accepts:       [3]float64{1.25, 2.25, 5.00}, // constant use behaves well, others are not great
+				rejects:       [3]float64{0.75, 0.75, 0.00},
+				cumulativeRPS: targetRPS,
+			}, run(t, roundSchedule(1, 5)))
+		})
```

Comment:
this is the problem

Reply:
I kind of understood the problem from the description, but I cannot understand it from the test description.